crates/core/src/table/file_format_options.rs

```diff
@@ -0,0 +1,308 @@
+        let state = sb.build();
+        return state;
+    }
+    state
+}
+
+#[cfg(feature = "datafusion")]
+fn build_writer_properties_tpo(
+    table_parquet_options: &Option<TableParquetOptions>,
+) -> Option<WriterProperties> {
+    use datafusion::common::file_options::parquet_writer::ParquetWriterOptions;
+    table_parquet_options.as_ref().map(|tpo| {
+        let mut tpo = tpo.clone();
+        tpo.global.skip_arrow_metadata = true;
+        ParquetWriterOptions::try_from(&tpo)
+            .expect("Failed to convert TableParquetOptions to ParquetWriterOptions")
+            .writer_options()
+            .clone()
+    })
+}
+
+#[cfg(feature = "datafusion")]
+fn build_writer_properties_factory_tpo(
+    table_parquet_options: &Option<TableParquetOptions>,
+) -> Option<Arc<dyn WriterPropertiesFactory>> {
+    let props = build_writer_properties_tpo(table_parquet_options);
+    props.map(|wp| {
+        Arc::new(SimpleWriterPropertiesFactory::new(wp)) as Arc<dyn WriterPropertiesFactory>
+    })
+}
+
+pub fn build_writer_properties_factory_wp(
+    writer_properties: WriterProperties,
+) -> Arc<dyn WriterPropertiesFactory> {
+    Arc::new(SimpleWriterPropertiesFactory::new(writer_properties))
+}
+
+pub fn build_writer_properties_factory_default() -> Arc<dyn WriterPropertiesFactory> {
+    Arc::new(SimpleWriterPropertiesFactory::default())
+}
+
+#[async_trait]
+pub trait WriterPropertiesFactory: Send + Sync + std::fmt::Debug + 'static {
+    fn compression(&self, column_path: &ColumnPath) -> Compression;
+    async fn create_writer_properties(
+        &self,
+        file_path: &Path,
+        file_schema: &Arc<ArrowSchema>,
+    ) -> DeltaResult<WriterProperties>;
+}
+
+#[derive(Clone, Debug)]
+pub struct SimpleWriterPropertiesFactory {
+    writer_properties: WriterProperties,
+}
+
+impl SimpleWriterPropertiesFactory {
+    pub fn new(writer_properties: WriterProperties) -> Self {
+        Self { writer_properties }
+    }
+}
+
+impl Default for SimpleWriterPropertiesFactory {
+    fn default() -> Self {
+        let writer_properties = WriterProperties::builder()
+            .set_compression(Compression::SNAPPY) // Code assumes Snappy by default
+            .set_created_by(format!("delta-rs version {}", crate_version()))
+            .build();
+        Self { writer_properties }
+    }
+}
+
+#[async_trait]
+impl WriterPropertiesFactory for SimpleWriterPropertiesFactory {
+    fn compression(&self, column_path: &ColumnPath) -> Compression {
+        self.writer_properties.compression(column_path)
+    }
+
+    async fn create_writer_properties(
+        &self,
+        file_path: &Path,
+        _file_schema: &Arc<ArrowSchema>,
+    ) -> DeltaResult<WriterProperties> {
+        info!("Called create_writer_properties for file: {file_path}");
+        Ok(self.writer_properties.clone())
+    }
+}
+
+// More advanced factory with KMS support
+#[cfg(feature = "datafusion")]
+#[derive(Clone, Debug)]
+pub struct KMSWriterPropertiesFactory {
```

Comment:
It might make sense to move things around a bit so this lives in the  `encryption` module alongside `TableEncryption`. `encryption` is also nested under `operations` but could maybe be next to `file_format_options` under the `table` module?

`KmsFileFormatOptions` is defined in the example, but maybe it should be in the same module as this type too.